docs/RFCS/20210527_changefeed_http_sink.md

```diff
@@ -0,0 +1,202 @@
+endpoints don't have to be health-checked and maintained throughout the 
+lifespan of the changefeed. So `Dial()` and `Close()`in the `Sink` interface 
+are no-ops in this scenario, at least for the sink itself (channel still needs 
+to be closed).
+
+The majority of the http connection logic should be wrapped in an 
+[`httputil.Client`](../../pkg/util/httputil/client.go) object (wrapper around Go's 
+[`http.Client`](https://golang.org/src/net/http/client.go#L57)) which will
+handle most of the TLS and other config setup mentioned below. The HTTP sink
+struct might look something like this:
+
+```
+type httpSink struct {
+  u sinkURL
+  client *httputil.Client
+	mu struct {
+		syncutil.Mutex
+		inflight int64
+		flushCh chan struct{}
+	}
+}
+```
+
+## Concurrency
+
+Each `POST` to the HTTP endpoint is executed asynchronously through calls to 
+`EmitRow()` and `EmitResolvedTimestamp()`. We use a channel here, `flushCh`,
+which we will send the http response object once it is received, or the error
+(if one occurs). We use a channel as opposed to something like a `WaitGroup` to
+receive a response (which may be needed for ack functionality later on) or an 
+error.
+
+```
+func (s *httpSink) EmitRow (...) error {
+  ...
+  res, err := s.client.Post(...)
+  s.mu.Lock()
+  s.mu.inflight++
+  s.mu.Unlock()
+  go func() {
+    s.mu.flushCh <- struct {
+      res *http.Response
+      err error
+    } {
+      res: res,
+      err: err,
+    }
+  }
+  ...
+  return err
+}
+```
+
+On the `Flush()` side, the the `mu` field can act as a semaphore, and allow
+reading from the channel until the sink has been flushed.
+
+```
+func (s *httpSink) Flush (...) error {
+  ...
+  s.mu.Lock()
+  defer s.mu.Unlock()
+  for s.mu.inflight > 0 {
+    select {
+      case ctx.Done():
+        return ctx.Err()
+      case flushed := <- s.mu.flushCh:
+        s.mu.inflight--
+        // can add additional ack logic here with flushed.res if needed
+        log.Infof("Sink returned response: %v", flushed.res)
+  }
+  ...
+}
+```
+
+## Configurability
+
+The HTTP sink should be configurable through a few different options and sink
+URL query params, a few of which are already supported by other sinks:
+
+* `ca_cert` - supply a custom CA cert for server verification
+* `insecure_tls_skip_verify` - used to disable client-side verification of CA
+  certs (corresponds to Go's 
+  [`tls.Config.InsecureSkipVerify`](https://golang.org/src/crypto/tls/common.go#L624) 
+  flag)
+
+These TLS-related query params should only be enabled if the provided sink URL
+is prefixed with `https`, throwing an error otherwise.
+
+Some HTTP-based options should also be added for `CREATE CHANGEFEED` statement:
+
+* `http_auth_header` - a user-supplied `Authorization` header for either basic 
+  base64-encoded auth or an OAuth token. As this contains credentials, it
+  should be redacted when appearing in the `description` column of `SHOW JOBS`/
+  `SHOW CHANGEFEED JOBS` etc.
```

Comment:
I think allowing people to specify an exact `Authorization` header is a good feature since a lot of users will be talking to something that uses long-lived bearer token-based auth system.

However, this point specifically mentions OAuth.  If it is then case that we want OAuth 2.0 support as a requirement for this feature, then we'll likely want some additional options specifically to support that it. In OAuth 2, the short-lived tokens sent in Authorization headers are obtained via some _other_ authentication flow.  My guess is that we'd want to start with supporting the client credentials flow where we will need a client_id and client_secret to be provided.

My guess is we could leave Oauth2 support out of the first version of this and still have a useful feature for many users.

Reply:
good point, let's specify the list as being basic auth + OAuth1 for now (we can discuss others if needed) with OAuth2 being a future item